lib/ecto/adapters/postgres.ex

```diff
@@ -79,66 +81,40 @@ defmodule Ecto.Adapters.Postgres do
       extra = extra <> " LC_CTYPE='#{lc_ctype}'"
     end
 
-    {output, status} =
-      run_with_psql opts, "CREATE DATABASE \"" <> database <>
-                          "\" ENCODING='#{encoding}'" <> extra
+    command = "CREATE DATABASE #{database} " <>
+              "ENCODING '#{encoding}'" <> extra
 
-    cond do
-      status == 0                       -> :ok
-      String.contains?(output, "42P04") -> {:error, :already_up}
-      true                              -> {:error, output}
+    case run_query(opts, command) do
+      {:ok, _result} ->
+        :ok
+      {:error, %Postgrex.Error{postgres: %{code: :duplicate_database}}} ->
+        {:error, :already_up}
+      error -> error
     end
   end
 
   @doc false
   def storage_down(opts) do
-    {output, status} = run_with_psql(opts, "DROP DATABASE \"#{opts[:database]}\"")
-
-    cond do
-      status == 0                       -> :ok
-      String.contains?(output, "3D000") -> {:error, :already_down}
-      true                              -> {:error, output}
+    command = "DROP DATABASE #{opts[:database]}"
+
+    case run_query(opts, command) do
+      {:ok, _result} ->
+        :ok
+      {:error, %Postgrex.Error{postgres: %{code: :invalid_catalog_name}}} ->
+        {:error, :already_down}
+      error -> error
     end
   end
 
-  defp run_with_psql(database, sql_command) do
-    unless System.find_executable("psql") do
-      raise "could not find executable `psql` in path, " <>
-            "please guarantee it is available before running ecto commands"
-    end
+  defp run_query(opts, sql) do
```

Comment:
How do you feel about making `run_query/2` return `{:error,  {code, message}}`? This will simplify the calling code to:
```elixir
case run_query(opts, command) do
  {:ok, _result} -> :ok
  {:error, {:duplicate_database, _message}} -> {:error, :already_up}
  {:error, {_, message}} -> {:error, message} # This would also be more like the original implementation
end
```